### PR TITLE
feat(totp): initial recovery codes

### LINKED ISSUE #2294: TOTP RecoveryCode API
#### GET /recoveryCodes

:lock: HAWK-authenticated with verified session token

Generate new TOTP recovery codes

##### Request body

* `{}`

##### Response body

* array of recovery codes `['123-123-123', '123-123-123']`

##### Error responses

Failing requests may be caused
by the following errors
(this is not an exhaustive list):

* `code: 400, errno: XXX`

### LINKED ISSUE #2310: TOTP Notification emails
from mtg:
- [x] Send an email when TOTP enabled
- [x] Send an email when TOTP disabled
- [x] New device sign-in with TOTP (reusing new device sign-in)
- [ ] New recovery codes generated
- [ ] Recovery code used for login

### PR DESCRIPTION
**Pull request by vbudhram**
Fixes #2294 
Fixes #2310

### CODE REVIEW

**Comment by vbudhram** [comment]
@mozilla/fxa-devs r?

To help with testing, you should be able to use https://github.com/mozilla/fxa-content-server/pull/5992.

File: config/index.js [hunk 1]
```diff
@@ -796,6 +796,18 @@ var conf = convict({
...
       format: 'nat',
       env: 'TOTP_STEP_SIZE'
+    },
+    recoveryCodes: {
+      codeLength: {
+        doc: 'The length of a recovery code',
+        default: 8,
+        env: 'RECOVERY_CODE_LENGTH'
+      },
+      codeCount: {
```
Comment (philbooth) [inline config/index.js]: No biggie, but as these are already nested inside a `recoveryCodes` parent node, seems like they could just be `count` and `length` without losing clarity.

File: docs/api.md [hunk 2]
```diff
@@ -2141,6 +2146,48 @@ will be returned.
...
+  <!--begin-response-body-get-recoverycodes-recoveryCodes-->
+  
+  <!--end-response-body-get-recoverycodes-recoveryCodes-->
+
+
+#### POST /session/verify/recoveryCode
+
+:lock: HAWK-authenticated with session token
+<!--begin-route-post-sessionverifyrecoverycode-->
+
```
Comment (philbooth) [inline docs/api.md]: I think we should try to add a sentence or two describing the new routes in the docs, even if their usage seems obvious to us from the names.

File: lib/customs.js [hunk 3]
```diff
@@ -16,8 +16,11 @@ module.exports = function (log, error) {
 
   // Perform a deep clone of payload and remove user password.
   function sanitizePayload(payload) {
+    if (! payload) {
+      return
+    }
     // Once we move to Node4, use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign
```
Comment (philbooth) [inline lib/customs.js]: Looks like this comment can be deleted now right?

File: lib/routes/recovery-codes.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    {
+      method: 'POST',
+      path: '/session/verify/recoveryCode',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            code: isA.string().min(RECOVERY_CODE_LENGTH).max(RECOVERY_CODE_LENGTH).regex(HEX_STRING).required()
```
Comment (philbooth) [inline lib/routes/recovery-codes.js]: If they're always going to be the same value, we could replace the `.min()` and `.max()` here with a single `.length()` call.

File: lib/routes/recovery-codes.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+        function verifySession() {
+          if (sessionToken.tokenVerificationId) {
+            return db.verifyTokensWithMethod(sessionToken.id, 'recovery-code')
+          }
+        }
+
+        function sendEmailNotification() {
+          if (! isValidCode) {
+            return
```
Comment (philbooth) [inline lib/routes/recovery-codes.js]: Is this unreachable code?

It looks like `isValidCode` will only be `false` if `consumeRecoveryCode` rejected. But if `consumeRecoveryCode` rejected, we'd have fallen straight through to the final `reply` handler. So it seems like `isValidCode` doesn't really serve any purpose in this endpoint, unless I'm reading the code wrong?

File: test/remote/recovery_code_tests.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+'use strict'
+
+const assert = require('insist')
+const config = require('../../config').getProperties()
+const TestServer = require('../test_server')
+const Client = require('../client')()
+const otplib = require('otplib')
+
+describe('remote totp', function () {
```
Comment (philbooth) [inline test/remote/recovery_code_tests.js]: Should this be `'remote recovery codes'`?

File: test/remote/recovery_code_tests.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+              .then((emailData) => {
+                assert.equal(emailData.headers['x-template-name'], 'postAddTwoStepAuthenticationEmail', 'correct template sent')
+              })
+          })
+      })
+  })
+
+  it('should create recovery codes', () => {
+    assert.ok(recoveryCodes)
+    assert.equal(recoveryCodes.length, 8, 'recovery codes returned')
```
Comment (philbooth) [inline test/remote/recovery_code_tests.js]: Should we also assert the content of `recoveryCodes`, e.g. that the strings are the expected length, only contain hex, whatever?

File: lib/senders/templates/post_consume_recovery_code.html [hunk 8]
```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
```
[the remark this answers is not in the document]
Comment (vbudhram) [reply lib/senders/templates/post_consume_recovery_code.html]: I keep missing these 😕, made https://github.com/mozilla/fxa-auth-server/issues/2360 to help better document steps.

**Comment by vbudhram** [comment]
@philbooth Updated!

File: docs/api.md (hunk 2 above)
Comment (philbooth) [reply docs/api.md]: ...I'm still not seeing a description for the new endpoints? 😄

File: lib/metrics/amplitude.js [hunk 9]
```diff
@@ -85,6 +85,8 @@ const EMAIL_TYPES = {
   postRemoveSecondaryEmail: 'secondary_email',
   postVerifyEmail: 'registration',
   postVerifySecondaryEmail: 'secondary_email',
+  postConsumeRecoveryCodeEmail: 'recovery_code_email',
+  postNewRecoveryCodesEmail: 'recovery_code_email',
```
Comment (philbooth) [inline lib/metrics/amplitude.js]: I'd tweak these, I don't think they want the `_email` suffix. And maybe they should more generic too: `totp` or `2fa` perhaps?

**Pull request merged**